EIPS/eip-5988.md

```diff
@@ -0,0 +1,163 @@
+
+## Motivation
+
+[Poseidon](../assets/eip-5988/papers/poseidon_paper.pdf) is an arithmetic hash function that is designed to be efficient for Zero-Knowledge Proof Systems.
+Ethereum adopts a rollup centric roadmap and hence must adopt facilities for L2s to be able to communicate with the EVM in an optimal manner.
+
+ZK-Rollups have particular needs for cryptographic hash functions that can allow for efficient verification of proofs.
+
+The Poseidon hash function is a set of permutations over a prime field, which makes it particularly well-suited for the purpose of building efficient ZK / Validity rollups on Ethereum.
+
+Poseidon is one of the most efficient hashing algorithms that can be used in this context.
+Moreover it is compatible with all major proof systems (SNARKs, STARKs, Bulletproofs, etc...).
+This makes it a good candidate for a precompile that can be used by many different ZK-Rollups.
+
+An important point to note is that ZK rollups using Poseidon have chosen different sets of parameters, which makes it harder to build a single precompile for all of them.
+
+However, we can still build a generic precompile that supports arbitrary parameters, and allow the ZK rollups to choose the parameters they want to use.
+
+This is the approach that we have taken in this EIP.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+### Parameters
+
+| Constant                      | Value |
+| ----------------------------- | ----- |
+| `FORK_BLKNUM`                 | `TBD` |
+| `GAS_COST`                    | `TBD` |
+| `POSEIDON_PRECOMPILE_ADDRESS` | `0xA` |
+
+Here are the Poseidon parameters that the precompile will support:
+
+- `p`: the prime field modulus.
+- `security_level`: the security level measured in bits. Denoted `M` in the Poseidon paper.
+- `alpha`: the power of S-box.
+- `input_rate`: the size of input.
+- `t`: the size of the state.
+- `full_round`: the number of full rounds. Denoted as `R_F` in the Poseidon paper.
+- `partial_round`: the number of partial rounds. Denoted as `R_P` in the Poseidon paper.
+
+The input to the precompile MUST be encoded as follows:
+
+- `p`: 32 bytes
+- `security_level`: 2 bytes
+- `alpha`: 1 byte
+- `input_rate`: 2 bytes
+- `t`: 1 byte
+- `full_round`: 1 byte
+- `partial_round`: 1 byte
+- `input`: `input_rate` \* 32 bytes
+
+The encoding of the precompile input is the following:
+
+```text
+[32 bytes for p][2 bytes for security_level][1 byte for alpha][2 bytes for input_rate][1 byte for t][1 byte for full_round][1 byte for partial_round][input_rate * 32 bytes for input]
+```
+
+The precompile should compute the hash function as [specified in the Poseidon paper](../assets/eip-5988/papers/poseidon_paper.pdf) and return hash output.
```

Comment:
Since "how the precompile contract computes a hash is computed" is the core part of this specification. I lean towards suggesting author to spell out algorithm rather than briefly mention "compute the hash function as in `that paper`".